File: shoop/notify/actions/order.py
# -*- coding: utf-8 -*-
# This file is part of Shoop.
#
# Copyright (c) 2012-2015, Shoop Ltd. All rights reserved.
#
# This source code is licensed under the AGPLv3 license found in the
# LICENSE file in the root directory of this source tree.
from shoop.core.models import Order
from shoop.notify.base import Action, Binding, ConstantUse
from shoop.notify.typology import Model, Text


class AddOrderLogEntry(Action):
    identifier = "add_order_log_entry"
    order = Binding("Order", Model("shoop.Order"), constant_use=ConstantUse.VARIABLE_ONLY)
    message = Binding("Message", Text, constant_use=ConstantUse.VARIABLE_OR_CONSTANT)
    message_identifier = Binding("Message Identifier", Text, constant_use=ConstantUse.VARIABLE_OR_CONSTANT)

    def execute(self, context):
        order = self.get_value(context, "order")
        if not order:  # pragma: no cover
            return
        message = self.get_value(context, "message")
        message_identifier = self.get_value(context, "message_identifier") or None
        assert isinstance(order, Order)
        order.add_log_entry(message=message, identifier=message_identifier)

File: shoop/admin/modules/orders/static_src/create/reducers/order.js
/**
 * This file is part of Shoop.
 *
 * Copyright (c) 2012-2016, Shoop Ltd. All rights reserved.
 *
 * This source code is licensed under the AGPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {handleActions} from "redux-actions";
import _ from "lodash";

function updateTotals(state, {payload}) {
    const updates = {};
    const {lines, methods} = payload();
    var total = 0;
    _.map(lines, (line) => {
        total += line.total;
    });
    updates.total = +total.toFixed(2);
    return _.assign({}, state, updates);
}

export default handleActions({
    beginCreatingOrder: ((state) => _.assign(state, {creating: true})),
    endCreatingOrder:  ((state) => _.assign(state, {creating: false})),
    updateTotals,
    setOrderSource: ((state, {payload}) => _.assign(state, {source: payload})),
    clearOrderSourceData: ((state) => _.assign(state, {source: null}))
}, {
    creating: false,
    source: null,
    total: 0
});
